walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,142 @@
 
 package com.google.sps;
 
+import com.google.appengine.repackaged.com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Gets the available time slot for new meeting given existing meeting events.
+   * @param events Existing meeting events, assuming all events end before EOD.
+   * @param request Request to set up a new meeting with duration not exceeding the whole day.
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty collection if invalid input.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long durationInMinutes = request.getDuration();
+    if (!isValidParams(events, durationInMinutes)) {
+      // return Collections.unmodifiableList(new ArrayList<>());
+      return ImmutableList.of();
+    }
+
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> attendeesWithOptional = new HashSet<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return ImmutableList.copyOf(Arrays.asList(TimeRange.WHOLE_DAY));
```

Comment:
`return ImmutableList.of(TimeRange.WHOLE_DAY);`

Reply:
Fixed